system/mp/mpi_wrapper.cpp

```diff
@@ -0,0 +1,526 @@
+    return replyTag;
+}
+
+hpcc_mpi::CommStatus hpcc_mpi::MPIComm::sendData(rank_t dstRank, mptag_t mptag, CMessageBuffer &mbuf, unsigned timeout)
+{
+    _TF("sendData", dstRank, mptag, mbuf.getReplyTag(), timeout);
+    mpiInitializedCheck();
+    CTimeMon tm(timeout);
+    unsigned remaining;
+    int target = getRank(dstRank); int tag = getMPITag(mptag);
+    bool timedout = false; bool error = false; bool canceled = false; bool completed;
+    bool bufferedSendComplete = false;
+
+    //TODO: find a better way to send the reply tag
+    pushMPTag(mbuf, mbuf.getReplyTag());
+    MPI_Request req;
+    while(!bufferedSendComplete && !canceled)
+    {
+        int error = MPI_Ibsend(mbuf.bufferBase(), mbuf.length(), MPI_BYTE, target, tag, this->get(), &req);
+        int errorClass;
+        MPI_Error_class(error, &errorClass);
+        if (errorClass == MPI_SUCCESS)
+            bufferedSendComplete = true;
+        else
+        {
+            if (errorClass == MPI_ERR_BUFFER)
+                usleep(WAIT_DELAY); //retry after giving some time for the buffers to clear up
+            else
+            {
+                throw errorClass;
+            }
+        }
+    }
+    popMPTag(mbuf);
+
+    timedout = tm.timedout(&remaining);
+
+    hpcc_mpi::CommStatus status =
+            error ? hpcc_mpi::CommStatus::ERROR
+                    : (timedout?    hpcc_mpi::CommStatus::TIMEDOUT
+                                    :canceled?  hpcc_mpi::CommStatus::CANCELED
+                                            :hpcc_mpi::CommStatus::SUCCESS);
+    return status;
+}
+
+hpcc_mpi::CommStatus hpcc_mpi::MPIComm::readData(rank_t &sourceRank, mptag_t &mptag, CMessageBuffer &mbuf, unsigned timeout)
+{
+    _TF("readData", sourceRank, mptag, timeout);
+    mpiInitializedCheck();
+    CTimeMon tm(timeout);
+    unsigned remaining;
+    bool incomingMessage = false; bool error = false; bool completed = false; bool canceled = false; bool timedout = false;
+
+    tm.timedout(&remaining);
+    int source = getRank(sourceRank);
+    int tag = getMPITag(mptag);
+
+    CommData* commData = new CommData(source, tag, *this);
+    addCommData(commData); //So that it can be cancelled from outside
+
+    MPI_Status stat = hasIncomingData(source, tag, *this, incomingMessage, error, canceled, remaining, commData);
+
+    timedout = tm.timedout(&remaining);
+    if (incomingMessage && !canceled)
+    {
+        int size;
+        MPI_Get_count(&stat,MPI_BYTE,&size);
+        assertex(size>0);
+        mbuf.setLength(size);
+        bool notCanceled = commData->lockFromCancellation();
+        if (notCanceled)
+        {
+            MPI_Irecv(mbuf.bufferBase(), mbuf.length(), MPI_BYTE, source, tag, this->get(), commData->request);
+            commData->releaseCancellationLock();
+        }
+        canceled = !notCanceled;
+        tm.timedout(&remaining);
+
+        MPI::Status stat = waitToComplete(completed, error, canceled, timedout, remaining, commData);
```

Comment:
Well it might be tricky. 
We cant do MPI_Irecv without MPI_Get_count (since we need to know how much space we need to reserve in the buffer passed for Irecv). And we cannot call MPI_Get_count without a MPI_Status object which tells about which incoming message to get the details from. And this MPI_Status is usually obtained via an MPI_Probe or MPI_Iprobe. 
So it would be hard to avoid the call for probing.